File: frontend/src/api/analyst.js
// Analyst API — SSE streaming from Cortex Agent
/**
 * Stream a natural-language query to the Cortex Agent.
 * Reads SSE events (delta, metadata, done, error) from the backend.
 */
export async function streamAnalystQuery(question, onChunk, onStatus) {
    const res = await fetch('/api/analyst/query', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ question }),
    });
    if (!res.ok) {
        const text = await res.text().catch(() => res.statusText);
        throw new Error(`HTTP ${res.status}: ${text}`);
    }
    // If the response is JSON (non-SPCS fallback), handle it directly
    const contentType = res.headers.get('content-type') ?? '';
    if (contentType.includes('application/json')) {
        const data = await res.json();
        if (data.answer)
            onChunk(data.answer);
        return undefined;
    }
    // SSE streaming
    const reader = res.body?.getReader();
    if (!reader)
        throw new Error('Response body is not readable');
    const decoder = new TextDecoder();
    let buffer = '';
    let metadata;
    try {
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            buffer += decoder.decode(value, { stream: true });
            // Process complete SSE events (separated by double newline)
            const parts = buffer.split('\n\n');
            buffer = parts.pop() ?? '';
            for (const part of parts) {
                if (!part.trim())
                    continue;
                let eventType = 'message';
                let data = '';
                for (const line of part.split('\n')) {
                    if (line.startsWith('event: ')) {
                        eventType = line.slice(7).trim();
                    }
                    else if (line.startsWith('data: ')) {
                        data = line.slice(6);
                    }
                }
                if (eventType === 'delta' && data) {
                    try {
                        const parsed = JSON.parse(data);
                        if (parsed.text)
                            onChunk(parsed.text);
                    }
                    catch {
                        // skip unparseable chunks
                    }
                }
                else if (eventType === 'status' && data) {
                    try {
                        const parsed = JSON.parse(data);
                        if (parsed.message && onStatus)
                            onStatus(parsed.message);
                    }
                    catch {
                        // skip
                    }
                }
                else if (eventType === 'metadata' && data) {
                    try {
                        metadata = JSON.parse(data);
                    }
                    catch {
                        // skip
                    }
                }
                else if (eventType === 'error' && data) {
                    try {
                        const parsed = JSON.parse(data);
                        onChunk(`\n\n[Error: ${parsed.message}]`);
                    }
                    catch {
                        onChunk('\n\n[Error: Unknown error]');
                    }
                }
                else if (eventType === 'done') {
                    break;
                }
            }
        }
    }
    finally {
        reader.releaseLock();
    }
    return metadata;
}

File: backend/app/routes/analyst.py
"""Cortex Agent route — natural-language queries over BDC data.

Uses the existing COACHING_AGENT (Cortex Agent) to answer natural-language
questions about calls, agents, dealer performance, etc.  The agent has
access to the BDC_ANALYTICS semantic view and call transcript search.

Streams SSE events to the browser for a real-time typing effect.
"""

import json
import sys

from fastapi import APIRouter
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from app.cortex import is_spcs, run_agent_stream_with_metadata

router = APIRouter()


def _log(msg: str) -> None:
    print(f"[ANALYST] {msg}", file=sys.stderr, flush=True)


class AnalystQuery(BaseModel):
    question: str


@router.post("/api/analyst/query")
async def analyst_query(req: AnalystQuery):
    """Stream an SSE response from the Cortex Agent for a NL analytics question."""

    if not is_spcs():
        # Non-SPCS fallback: return a single JSON response
        return {
            "answer": "Cortex Agent is only available when deployed to SPCS.",
            "sql": None,
            "results": [],
            "columns": [],
        }

    _log(f"Query: {req.question}")

    prompt = (
        "You are a BDC analytics assistant. Answer the following question about "
        "automotive BDC (Business Development Center) data using the tools available "
        "to you. Use the semantic view to write and execute SQL when needed. "
        "Be concise and data-driven in your response.\n\n"
        f"Question: {req.question}"
    )

    messages = [
        {"role": "user", "content": [{"type": "text", "text": prompt}]}
    ]

    async def _sse_generator():
        """Yield SSE events: delta (text chunks), metadata (final), done."""
        try:
            async for item in run_agent_stream_with_metadata(messages):
                if isinstance(item, dict) and item.get("__status__"):
                    yield f"event: status\ndata: {json.dumps({'message': item['message']})}\n\n"
                elif isinstance(item, dict):
                    # Final metadata item — strip internal flag
                    meta = {k: v for k, v in item.items() if k != "__metadata__"}
                    _log(f"Metadata: {meta}")
                    yield f"event: metadata\ndata: {json.dumps(meta)}\n\n"
                else:
                    yield f"event: delta\ndata: {json.dumps({'text': item})}\n\n"
        except Exception as e:
            _log(f"Stream error: {e}")
            yield f"event: error\ndata: {json.dumps({'message': str(e)})}\n\n"
        yield "event: done\ndata: [DONE]\n\n"

    return StreamingResponse(
        _sse_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )
